File: 2542-maximum-subsequence-score/2542-maximum-subsequence-score.py
class Solution:
    def maxScore(self, A, B, k):
        total = res = 0
        h = []
        for a,b in sorted(list(zip(A, B)), key=lambda ab: -ab[1]):
            heappush(h, a)
            total += a
            if len(h) > k:
                total -= heappop(h)
            if len(h) == k:
                res = max(res, total * b)
        return res

File: 2542-maximum-subsequence-score/2542-maximum-subsequence-score.cpp
class Solution {
public:
    long long maxScore(vector<int>& nums1, vector<int>& nums2, int k) {
        int n = nums1.size(); 
        vector<pair<int, int>> two; 
        for (int i = 0; i < n; ++i) 
            two.emplace_back(nums2[i], nums1[i]); 
        sort(two.begin(), two.end(), greater<>()); 
        for (int i = 0; i < n; ++i) {
            nums1[i] = two[i].second; 
            nums2[i] = two[i].first; 
        }
        priority_queue<int, vector<int>, greater<>> pq; 
        long long ans = 0, sum = 0; 
        for (int i = 0; i < n; ++i) {
            sum += nums1[i]; 
            pq.push(nums1[i]); 
            if (i >= k) {
                sum -= pq.top(); pq.pop(); 
            }
            if (i >= k-1) ans = max(ans, sum * nums2[i]); 
        }
        return ans; 
    }
};